crypto/ml_dsa/ml_dsa_key.c

```diff
@@ -0,0 +1,436 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/core_dispatch.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <openssl/proverr.h>
+#include <openssl/rand.h>
+#include "ml_dsa_local.h"
+#include "ml_dsa_key.h"
+#include "ml_dsa_params.h"
+#include "ml_dsa_matrix.h"
+#include "ml_dsa_hash.h"
+
+/**
+ * @brief Create a new ML_DSA_KEY object
+ *
+ * @param libctx A OSSL_LIB_CTX object used for fetching algorithms.
+ * @param propq The property query used for fetching algorithms
+ * @param alg The algorithm name associated with the key type
+ * @returns The new ML_DSA_KEY object on success, or NULL on malloc failure
+ */
+ML_DSA_KEY *ossl_ml_dsa_key_new(OSSL_LIB_CTX *libctx, const char *propq,
+                                const char *alg)
+{
+    ML_DSA_KEY *ret;
+    const ML_DSA_PARAMS *params = ossl_ml_dsa_params_get(alg);
+
+    if (params == NULL)
+        return NULL;
+
+    ret = OPENSSL_zalloc(sizeof(*ret));
+    if (ret != NULL) {
+        ret->libctx = libctx;
+        ret->params = params;
+        ret->shake128_md = EVP_MD_fetch(libctx, "SHAKE-128", propq);
+        ret->shake256_md = EVP_MD_fetch(libctx, "SHAKE-256", propq);
+        if (ret->shake128_md == NULL || ret->shake256_md == NULL)
+            goto err;
+    }
+    return ret;
+err:
+    ossl_ml_dsa_key_free(ret);
+    return NULL;
+}
+
+int ossl_ml_dsa_key_pub_alloc(ML_DSA_KEY *key)
+{
+    if (key->t1.poly != NULL)
+        return 1;
+    return vector_alloc(&key->t1, key->params->k);
+}
+
+int ossl_ml_dsa_key_priv_alloc(ML_DSA_KEY *key)
+{
+    size_t k = key->params->k, l = key->params->l;
+    POLY *poly;
+
+    if (key->s1.poly != NULL)
+        return 1;
```

Comment:
Return 0 here, if the key is already populated, mutation is not supported.

Reply:
This is allocation logic, this will not cause a mutation.

Reply:
But it hides the fact that the data is already allocated, and *now* should not be changed.